### ISSUE TITLE
Factor is exactly singular

### ISSUE THREAD

**Issue by YanhaoZhang** [lucmos/spectraldifferentiation#1]
Hi @lucmos Thanks a lot for your recommendation for this code. I find it is very helpful for me. May I ask you one quick question. Sometimes, I got the error message of 'Factor is exactly singular'. I check the matrix `Ascipy` in `Eigsh_torch`, it contains some really large numbers (around e^30). But when I save the vertices and faces as a mat file and run it again, this issue disappeared. May I know if you could you give some suggestions for this? Thanks a lot.

**Comment by Flegyas** [comment]
Some time ago I had the same problem and I discovered that it was caused by trying to process not watertight meshes. Could this be the case too?

**Comment by YanhaoZhang** [comment]
Hi @Flegyas, thanks for your prompt reply. I tested the code with a watertight mesh, but still got the error message. One thing I got quite confused is that, it only happens when I calculate the spectrum within my whole network code, but disappears if I save the vertices and faces and calculate the spectrum separately. 

I check the code again. I guess it might be because of the following code in `calc_LB_FEM_torch`. I find that sometimes, the angles contain nan or inf elements. But I am not very familiar with the algorithm, and do not know how to modify it. 

``` python 
        o = torch.mul(ab, ac)
        o = torch.sum(o, dim=1)
        o = torch.acos(o)
        o = torch.div(torch.cos(o), torch.sin(o))
        angles[i] = o
```

**Comment by YanhaoZhang** [comment]
Hi @lucmos I solved this issue by adding the following line before calculating the torch.acos. But I am still a bit confused on why this is happening as my mesh is generated using Poisson remeshing from meshlab. 

``` python
eps=1e-7
o = torch.clamp(o, min=-1+eps,max=1-eps)
```

Besides, may I ask you two more questions:
1. Whether a normalised mesh (vertex range between [-1, 1]) will get a better result or should I use the original mesh e.g., vertex range between [600, 1000]?
2. When calculating the decomposition, why only use the stiff matrix. I think the LB operator also consider the area right? 
Thanks a lot. :)

**Comment by lucmos** [comment]
Hello,
Usually, the 'Factor is exactly singular' happens when you have a broken mesh, flipped triangles, degenerate triangles, or triangles with a very small area close to zero. I think the differences you are seeing in the different executions are due to numerical errors and the behavior of the solver. Your clamping seems to be a workaround to triangles with ~0 areas.


Regarding the two other questions:

- Not sure which is the best option here. Keep in mind that if you use floats the numerical approximation error grows with big numbers
- That's just how the Cholesky decomposition  $B=LL^T$ works, you can reduce the generalized eigenvalue problem  $A z = \lambda B z$ to an equivalent normal eigenvalue problem $Cy = \lambda y$, with $C=L^{-1}AL^{-T}$. You can find more information here https://www.netlib.org/lapack/lug/node54.html

**Comment by YanhaoZhang** [comment]
Thanks a lot @lucmos! Now I understand it much better. Really appreciate your prompt reply. :)

**Comment by lucmos** [comment]
You're welcome! I am closing this, feel free to reopen if you have further questions!
